src/main/java/com/minecolonies/coremod/entity/ai/citizen/enchanter/EntityAIWorkEnchanter.java

```diff
@@ -228,36 +230,43 @@ private IAIState enchant()
             return getState();
         }
 
-        final int slot = InventoryUtils.findFirstSlotInItemHandlerWith(worker.getInventoryCitizen(), IS_ANCIENT_TOME);
-        if (slot != -1)
+        final ICitizenData data = worker.getCitizenData();
+        if (data != null)
         {
-            final ICitizenData data = worker.getCitizenData();
-            if (data != null)
-            {
-                final int openSlot = InventoryUtils.getFirstOpenSlotFromItemHandler(worker.getInventoryCitizen());
-                if (openSlot == -1)
-                {
-                    //Dump if there is no open slot.
-                    incrementActionsDone();
-                    progressTicks = 0;
-                    return IDLE;
-                }
+            List<RecordingItemHandler> handlers = getOwnBuilding().getHandlers().stream()
```

Comment:
using a RecordingItemHandler seems like quite the hacky workaround for the fullfillrecipe not giving back the required information, how about having a fullfillment which returns more information instead?

Reply:
That's certainly another valid option, although it has its own share of awkwardness by requiring changing the signature of a **lot** of existing code, and of having multiple output values from a single method (which are certainly doable, but is uglier).

I'm happy to change it to do that if there's a general preference for that approach, but only in that case.

Reply:
I guess we also could have an alternate which provides more information on return instead? Or doesnt do the full craft and insert and instead just consume & return stacks

Reply:
I think an alternative one that gives back a "CraftResult" object with the existing return value + the additional return is probably the best way.